### PR TITLE
Expose AmbiguousImplementationException and add tests

### PR DESCRIPTION
**Pull request by maryamariyan**
Exposes AmbiguousImplementationException and adds tests.
Fixes: #34124

cc: @danmosemsft @MichalStrehovsky

### CODE REVIEW

File: src/System.Runtime/ref/System.Runtime.cs [hunk 1]
```diff
@@ -6424,6 +6424,14 @@ public abstract partial class TypeInfo : System.Type, System.Reflection.IReflect
 }
 namespace System.Runtime
 {
+    public sealed class AmbiguousImplementationException : System.Exception
+    {
+        public AmbiguousImplementationException() { }
```
Comment (danmosemsft) [inline src/System.Runtime/ref/System.Runtime.cs]: please format like the other types in this file (ie no blank lines)

**Comment by danmosemsft** [comment]
Hm, it looks like BinaryFormatterTestData even includes simple exceptions with no fields, that just rely on the `[Serializable]` attribute. Can you please add this one there too?

**Comment by danmosemsft** [comment]
Does System.Runtime need to be added to src\System.Runtime.Serialization.Formatters\tests\BinaryFormatterTests.rd.xml , for UWP?

The failures here are not caused by that though
```
System.Runtime.Serialization.SerializationException : Unable to load type System.Runtime.AmbiguousImplementationException required for deserialization.
```

I am not sure why that is happening. Maybe @ViktorHofer has an idea.

**Comment by ViktorHofer** [comment]
Yep, @maryamariyan already reached out to me offline. We are looking into it later today. My guess is that the partial facade type forward doesn't work as it should / bits are not build correctly. Will give an update later.

**Comment by MichalStrehovsky** [comment]
This could be because Project N build doesn't know this type is public so it ends up reflection-blocking it. This would be a point in time issue until we switch N to 3.0 framework.

You can check if it's the case by doing a typeof on the type and accessing the FullName property. If you get guid garbage instead of the name, the type is blocked and there's nothing that can be done about it for now.

**Comment by danmosemsft** [comment]
@MichalStrehovsky that woudl be handled by the rd.xml edit I mentioned wouldn't it?

**Comment by maryamariyan** [comment]
> You can check if it's the case by doing a typeof on the type and accessing the FullName property. If you get guid garbage instead of the name, the type is blocked and there's nothing that can be done about it for now

@MichalStrehovsky I get the proper name back when accessing FullName property.

**Comment by MichalStrehovsky** [comment]
If something is blocked, it cannot be reenabled with rd.xml - blocking always wins. Rd.xml only helps if it's simply something we didn't generate.

Seems like it's not the case of blocking here though, although I'm a bit surprised. I'm out on vacation and can't help much more from a phone.

**Comment by jkotas** [comment]
My guess is that this is because of this hack: https://github.com/dotnet/corefx/blob/master/src/System.Runtime.Serialization.Formatters/src/System/Runtime/Serialization/Formatters/Binary/Converter.cs#L34

Binary serialization assumes that all types that live in `System.Private.CoreLib` should be serialized as if they lived in `mscorlib.dll` and expects a forwarder for them in `mscorlib.dll`.

Add a forwarder for this type to `mscorlib.dll` to make this hack work.

**Comment by ViktorHofer** [comment]
Lol Jan you were faster posting this. I came to the same conclusion. It seems this new exception type is the only type in S.P.CoreLib that doesn't need a type forward. We could either apply the type forward or try to fix the loading strategy for simple types.

**Comment by ViktorHofer** [comment]
`mscorlib.forward.cs` is mostly for internal types where TypeForwardedFrom doesn't work so I would have assumed you add the type forward in S.P.CoreLib but I'm ok with this if this unblocks you.

**Comment by jkotas** [comment]
Hmm, that's a good point. We should mark the type with `[System.Runtime.CompilerServices.TypeForwardedFrom("System.Runtime, ...` in CoreLib. And then it should just work. No need to touch mscorlib at all.

**Comment by ViktorHofer** [comment]
Great to see that it worked out!

File: src/System.Runtime.Serialization.Formatters/tests/BinaryFormatterTestData.cs [hunk 2]
```diff
@@ -653,6 +653,8 @@ private static IEnumerable<object[]> SerializableObjects()
...
 
 #if netcoreapp
+            var ambiguousImplementationException = new AmbiguousImplementationException();
```
Comment (ViktorHofer) [inline src/System.Runtime.Serialization.Formatters/tests/BinaryFormatterTestData.cs]: We should have passed a message and base exception to it as in other examples but I don't think we need to change it right now.

**Pull request merged**